as/plugins/org.jboss.ide.eclipse.as.wtp.core/src/org/jboss/ide/eclipse/as/wtp/core/server/publish/BinaryModulePublishRunner.java

```diff
@@ -0,0 +1,218 @@
+import org.eclipse.wst.server.core.model.IModuleResource;
+import org.eclipse.wst.server.core.model.IModuleResourceDelta;
+import org.jboss.ide.eclipse.as.core.util.ModuleResourceUtil;
+import org.jboss.ide.eclipse.as.wtp.core.ASWTPToolsPlugin;
+import org.jboss.ide.eclipse.as.wtp.core.server.behavior.IFilesystemController;
+
+public class BinaryModulePublishRunner {
+
+	private IModule[] module;
+	private IPath archiveDestination;
+	private IPath archiveDestinationWithName;
+	private IFilesystemController fc;
+	private IModuleResource[] all;
+	private IModuleResourceDelta[] delta;
+	public BinaryModulePublishRunner(IModule[] module, 
+			IPath archiveDestination, 
+			IPath archiveDestinationWithName, 
+			IFilesystemController fc,
+			IModuleResource[] all, IModuleResourceDelta[] delta) {
+		super();
+		this.module = module;
+		this.archiveDestination = archiveDestination;
+		this.archiveDestinationWithName = archiveDestinationWithName;
+		this.fc = fc;
+		this.all = all;
+		this.delta = delta;
+	}
+	/*
+	 * Use the delta to find all deleted resources. 
+	 * Then use the members to delete all those, too. 
+	 * @return
+	 */
+	public MultiStatus removeBinaryModule() throws CoreException {
+		MultiStatus ms = removeDeletedBinaryModule();
+		if( !isNestedSingleResourceBinaryModule()) {
+			removeMembers(all, ms);
+		}
+		return ms;
+	}
+	
+	private IStatus removeMembers(IModuleResource[] all, MultiStatus ms) throws CoreException {
+		for( int i = 0; i < all.length; i++ ) {
+			if( all[i] instanceof IModuleFolder) {
+				IModuleResource[] children = ((IModuleFolder)all[i]).members();
+				removeMembers(children, ms);
+			}
+			IPath relative = all[i].getModuleRelativePath();
+			IPath published = archiveDestination.append(relative).append(all[i].getName());
+			IStatus s = fc.deleteResource(published, new NullProgressMonitor());
+			if( s != null && !s.isOK()) {
+				ms.add(s);
+			}
+		}
+		return ms;
+	}
+	
+	
+	private MultiStatus removeNestedSingleResourceBinary(MultiStatus ms) throws CoreException {
+		IPath published = archiveDestinationWithName;
+		IStatus s = fc.deleteResource(published, new NullProgressMonitor());
+		if( s != null && !s.isOK()) {
+			ms.add(s);
+		}
+		return ms;
+	}
+	/*
+	 * Use the delta because the module is gone and has no members
+	 */
+	public MultiStatus removeDeletedBinaryModule() throws CoreException {
+		MultiStatus ms = new MultiStatus(ASWTPToolsPlugin.PLUGIN_ID, 0, "Errors while deleting binary module " + module[module.length-1].getName(), null);
+		IModuleResource[] deleted = getAllResources(delta, IModuleResourceDelta.REMOVED, true);
+		if( isNestedSingleResourceBinaryModule()) {
+			removeNestedSingleResourceBinary(ms);
+			return ms;
+		}
+		
+		for( int i = 0; i < deleted.length; i++ ) {
+			IPath relative = deleted[i].getModuleRelativePath();
+			IPath published = archiveDestination.append(relative).append(deleted[i].getName());
+			IStatus s = fc.deleteResource(published, new NullProgressMonitor());
```

Comment:
it might be worthy to pass a ProgressMonitor from calling methods to provide better progress reporting to user

Reply:
please see proposal https://github.com/robstryker/jbosstools-server/pull/18 for this last progress monitor not reused